core/src/main/java/org/pentaho/di/core/row/value/ValueMetaBase.java

```diff
@@ -4953,6 +4963,242 @@ protected void getOriginalColumnMetadata( ValueMetaInterface v, ResultSetMetaDat
+          }
+          break;
+
+        case java.sql.Types.INTEGER:
+          valtype = ValueMetaInterface.TYPE_INTEGER;
+          precision = 0; // Max 2.147.483.647
+          length = 9;
+          break;
+
+        case java.sql.Types.SMALLINT:
+          valtype = ValueMetaInterface.TYPE_INTEGER;
+          precision = 0; // Max 32.767
+          length = 4;
+          break;
+
+        case java.sql.Types.TINYINT:
+          valtype = ValueMetaInterface.TYPE_INTEGER;
+          precision = 0; // Max 127
+          length = 2;
+          break;
+
+        case java.sql.Types.DECIMAL:
+        case java.sql.Types.DOUBLE:
+        case java.sql.Types.FLOAT:
+        case java.sql.Types.REAL:
+        case java.sql.Types.NUMERIC:
+          valtype = ValueMetaInterface.TYPE_NUMBER;
+          length = originalPrecision;
+          precision = originalScale;
+          if ( length >= 126 ) {
+            length = -1;
+          }
+          if ( precision >= 126 ) {
+            precision = -1;
+          }
+
+          if ( originalColumnType == java.sql.Types.DOUBLE || originalColumnType == java.sql.Types.FLOAT || originalColumnType == java.sql.Types.REAL ) {
+            if ( precision == 0 ) {
+              precision = -1; // precision is obviously incorrect if the type if
+              // Double/Float/Real
+            }
+
+            // If we're dealing with PostgreSQL and double precision types
+            if ( databaseMeta.getDatabaseInterface() instanceof PostgreSQLDatabaseMeta && originalColumnType == java.sql.Types.DOUBLE
+                && precision >= 16 && length >= 16 ) {
+              precision = -1;
+              length = -1;
+            }
+
+            // MySQL: max resolution is double precision floating point (double)
+            // The (12,31) that is given back is not correct
+            if ( databaseMeta.getDatabaseInterface().isMySQLVariant() ) {
+              if ( precision >= length ) {
+                precision = -1;
+                length = -1;
+              }
+            }
+
+            // if the length or precision needs a BIGNUMBER
+            if ( length > 15 || precision > 15 ) {
+              valtype = ValueMetaInterface.TYPE_BIGNUMBER;
+            }
+          } else {
+            if ( precision == 0 ) {
+              if ( length <= 18 && length > 0 ) { // Among others Oracle is affected
+                // here.
+                valtype = ValueMetaInterface.TYPE_INTEGER; // Long can hold up to 18
+                // significant digits
+              } else if ( length > 18 ) {
+                valtype = ValueMetaInterface.TYPE_BIGNUMBER;
+              }
+            } else { // we have a precision: keep NUMBER or change to BIGNUMBER?
+              if ( length > 15 || precision > 15 ) {
+                valtype = ValueMetaInterface.TYPE_BIGNUMBER;
+              }
+            }
+          }
+
+          if ( databaseMeta.getDatabaseInterface() instanceof PostgreSQLDatabaseMeta
```

Comment:
Is there any way we can put this kind of stuff into the specific dialect ? I know this class is littered with these things already, but I'd rather not add more if we can avoid it.